examples/raytrace.dx

```diff
@@ -180,143 +127,231 @@ def sampleCosineWeightedHemisphere (k:Key) (normal: Vec3) : Vec3 =
+  case raymarch scene ray of
+    HitLight intensity -> intensity
+    HitNothing -> zero
+    HitObj _ _ -> zero
+
+def sampleSquare (hw:Real) (k:Key) : Position =
+ (kx, kz) = splitKey k
+ x = randuniform (- hw) hw kx
+ z = randuniform (- hw) hw kz
+ [x, 0.0, z]
+
+def sampleLightRadiance
+      (scene:Scene n) (osurf:OrientedSurface) (inRay:Ray) (k:Key) : Radiance =
+  (surfNor, surf) = osurf
+  (rayPos, _) = inRay
+  (MkScene objs) = scene
+  snd $ withAccum \radiance.
+    for i. case objs.i of
+      PassiveObject _ _ -> ()
+      Light lightPos hw _ ->
+        (dirToLight, distToLight) = directionAndLength $
+                                      lightPos + sampleSquare hw k - rayPos
+        case positiveProjection dirToLight surfNor of
+          False -> () -- light on the far side of current surface
+          True ->
+            fracSolidAngle = (relu $ dot dirToLight yHat) * sq hw / (pi * sq distToLight)
+            outRay = (rayPos, dirToLight)
+            coeff = fracSolidAngle * probReflection osurf inRay outRay
+            radiance += coeff .* rayDirectRadiance scene outRay
+
+def trace (params:Params) (scene:Scene n) (init_ray:Ray) (k:Key) : Color =
+  (_, max_bounces) = params
+  -- TODO: we ought to be able to use an accumulator here, but there's a bug
+  noFilter = [1.0, 1.0, 1.0]
+  iter (noFilter, zero, init_ray) $
+    \i (filter, radiance, ray).
+      case i >= max_bounces of
+        True -> Done radiance
+        False -> case raymarch scene ray of
+          HitNothing -> Done radiance
+          HitLight intensity -> case i == 0 of
+            True  -> Done intensity  -- TODO: scale etc
+            False -> Done radiance
+          HitObj incidentRay osurf ->
+            (k1, k2) = splitKey $ hash k i
+            lightRadiance    = sampleLightRadiance scene osurf incidentRay k1
+            outRayHemisphere = sampleReflection          osurf incidentRay k2
+            newFilter = surfaceFilter filter (snd osurf)
+            newRadiance = radiance + applyFilter newFilter lightRadiance
+            Continue (newFilter, newRadiance, outRayHemisphere)
+
+def avgRayColor (params:Params) (scene:Scene m) (ray:Ray) (k:Key) : Color =
+  (num_samples, _) = params
+  sampleAveraged (trace params scene ray) num_samples k
+
+-- TODO: add number of pixels once we can hide sizes
+-- sensor half-width, pinhole-sensor distance, pinhole position
+-- (Assumes we're looking towards -z.)
+Camera = (Position & Real & Real)
+
+def cameraRays (n:Int) (camera:Camera) : Fin n => Fin n => Ray =
+  -- images indexed from top-left
+  (pos, halfWidth, sensorDist) = camera
+  ys = reverse $ linspace (Fin n) (neg halfWidth) halfWidth
+  xs =           linspace (Fin n) (neg halfWidth) halfWidth
+  for i j. (pos, normalize [xs.j, ys.i, neg sensorDist])
+
+def takePicture
+       (params:Params) (scene:Scene m) (n:Int) (camera:Camera)
+      : ColorImage n n =
+  rays = cameraRays n camera
```

Comment:
Would it make sense to randomly sample the camera ray within the square for the output pixel, instead of always using the exact center? Then I think you'd get antialiasing for object boundaries automatically.

Reply:
Good thought! It's like applying a low-pass analog filter at the Nyquist limit before digitizing. Since we're not limited by analog circuitry, we could apply a perfect low-pass filter. Short of that, here's what it looks like if we apply a square blur to the rays leaving the camera. It smooths out the edges of the sphere quite nicely.

![rendering4](https://user-images.githubusercontent.com/8833761/89574065-8d8edc80-d7f9-11ea-9c0b-f59724e49cc6.png)